app/server.py

```diff
@@ -0,0 +1,101 @@
+import asyncio
+import signal
+import sys
+import aiohttp
+import yaml
+from os.path import join, dirname
+from aiohttp import web
+
+routes = web.RouteTableDef()
+PROJECT_ROOT = dirname(dirname(__file__))
+jobs = []
+
+
+async def handle_index(_):
+    with open(join(PROJECT_ROOT, 'app', 'index.html'), 'r') as f:
+        return web.Response(body=f.read(), content_type='text/html')
+
+
+async def websocket_handler(request):
+    ws = web.WebSocketResponse()
+    await ws.prepare(request)
+
+    async for msg in ws:
+        if msg.type == aiohttp.WSMsgType.TEXT:
+            if msg.data == 'close':
+                await ws.close()
+            else:
+                p = await asyncio.create_subprocess_shell(
+                    msg.data,
```

Comment:
This spawns a process based on data received from an unauthenticated websocket.

Which basically means... an attacker can prepare a website that would connect to localhost websocket (`new WebSocket("ws://127.0.0.1:8080/ws");`) and then send proper payload to execute arbitrary code on victim's machine.

It would be much better not to spawn any processes at all and if it is really needed, it must be authenticated.

Reply:
See also:
- https://medium.com/@sharan.panegav/account-takeover-using-cross-site-websocket-hijacking-cswh-99cf9cea6c50
- https://security.stackexchange.com/questions/76816/preventing-csrf-attacks-against-websocket-communications

Reply:
@disconnect3d does Ansible provides API for launching playbooks other than from command line?

Reply:
@summerisgone Yes, take a look at this https://docs.ansible.com/ansible/latest/dev_guide/developing_api.html

Reply:
Thanks, I probably missed that API from the first time. Will check it out!

Reply:
@jackivanov I'm not good at ansible, could you help me or give an advice?
I'm able to run playbook from this snippet:
```python
PlaybookCLI(['ansible-playbook', 'main.yml']).run()
```
But in that case I can't display feedback, which is valuable, since installation process takes a long time and may exit with error. I'm looking for snippet which allows to set up custom `stdout_callback` and play scenario from yaml file, like above.

Reply:
@summerisgone You don't seem to be running ansible via the API.  You can find a working example by the link I sent above

Reply:
@jackivanov example you provided doesn't allow to load playbooks

Reply:
I'm not well familiar with Ansible API, but as far I recall, there is a class called `PlaybookExecutor` that can be used. See [here](https://github.com/ansible/ansible/blob/stable-2.8/lib/ansible/executor/playbook_executor.py) and [here](https://github.com/ansible/ansible/blob/stable-2.8/test/units/executor/test_playbook_executor.py)

Reply:
@jackivanov or @disconnect3d could you please take another look on current approach of executing playbooks from web?